webapp/channels/src/utils/desktopAPI.ts

```diff
@@ -0,0 +1,288 @@
+     */
+    private postMessageListener = ({origin, data: {type, message}}: {origin: string; data: {type: string; message: unknown}}) => {
+        if (origin !== window.location.origin) {
+            return;
+        }
+
+        const listeners = this.postMessageListeners?.get(type);
+        if (!listeners) {
+            return;
+        }
+
+        listeners.forEach((listener) => {
+            listener(message);
+        });
+    };
+
+    /**
+     * @deprecated
+     */
+    addPostMessageListener = (channel: string, listener: (message: any) => void) => {
+        if (this.postMessageListeners?.has(channel)) {
+            this.postMessageListeners.set(channel, this.postMessageListeners.get(channel)!.add(listener));
+        } else {
+            this.postMessageListeners?.set(channel, new Set([listener]));
+        }
+    };
+
+    /**
+     * @deprecated
+     */
+    removePostMessageListener = (channel: string, listener: (message: any) => void) => {
+        const set = this.postMessageListeners?.get(channel);
+        set?.delete(listener);
+        if (set?.size) {
+            this.postMessageListeners?.set(channel, set);
+        } else {
+            this.postMessageListeners?.delete(channel);
+        }
+    };
+
+    /**
+     * @deprecated
+     */
+    invokeWithMessaging = <T, T2>(
+        sendChannel: string,
+        sendData?: T,
+        receiveChannel?: string,
+    ) => {
+        return new Promise<T2>((resolve) => {
+            /* create and register a temporary listener if necessary */
+            const response = ({origin, data: {type, message}}: {origin: string; data: {type: string; message: T2}}) => {
+                /* ignore messages from other frames */
+                if (origin !== window.location.origin) {
+                    return;
+                }
+
+                /* ignore messages from other channels */
+                if (type !== (receiveChannel ?? sendChannel)) {
+                    return;
+                }
+
+                /* clean up listener and resolve */
+                window.removeEventListener('message', response);
+                resolve(message);
+            };
+
+            window.addEventListener('message', response);
+            window.postMessage(
+                {type: sendChannel, message: sendData},
+                window.location.origin,
+            );
+        });
+    };
+}
+
+const desktopApp = new DesktopApp();
+export default desktopApp;
+
+/*******************************
+ * API Methods
+ *******************************/
+
+/**
+ * Invokes
+ */
+
+export const getBrowserHistoryStatus = () => {
+    if (window.desktopAPI?.requestBrowserHistoryStatus) {
+        return window.desktopAPI.requestBrowserHistoryStatus();
+    }
+
+    return desktopApp.invokeWithMessaging<void, {enableBack: boolean; enableForward: boolean}>(
+        'history-button',
+        undefined,
+        'history-button-return',
+    ).then(({enableBack, enableForward}) => {
```

Comment:
Again, I prefer the await nomenclature than the `then` one, but 0/5.

Reply:
Thanks for catching these :)